kernel/src/hil/public_key_crypto/keys.rs

```diff
@@ -283,3 +283,110 @@ pub trait RsaPrivKeyMut: PubPrivKeyMut + RsaKeyMut {
+/// representation of keys. They may be stored in memory, flash, or in secure
+/// element (where the actual key may not be accessible). Users of this
+/// interface can select which key is active for cryptographic operations. There
+/// is no assumption for implementers of this interface that keys can be added
+/// or changed or that keys can be specified by their actual bytes in a slice.
+///
+/// Selecting a key is asynchronous as it may require communication over a bus
+/// or waiting for an interrupt.
+///
+/// Keys are specified by an index starting at zero and going to
+/// `get_key_count()-1`.
+///
+/// The stored keys can be public or private keys.
+pub trait SelectKey<'a> {
+    /// Return the number of keys that the device can switch among.
+    ///
+    /// Each key must be identifiable by a consistent index.
+    ///
+    /// This operation is asynchronous and its completion is signaled by
+    /// `get_key_count_done()`.
+    ///
+    /// ## Return
+    ///
+    /// `Ok()` if getting the count has started. Otherwise:
+    /// - `Err(ErrorCode::FAIL)` if the key count could not be started and there
+    ///   will be no callback.
+    fn get_key_count(&self) -> Result<(), ErrorCode>;
+
+    /// Set the key identified by its index as the active key.
+    ///
+    /// Indices start at 0 and go to `get_key_count() - 1`.
+    ///
+    /// This operation is asynchronous and its completion is signaled by
+    /// `select_key_done()`.
+    ///
+    /// ## Return
+    ///
+    /// `Ok()` if the key select operation was accepted. Otherwise:
+    /// - `Err(ErrorCode::INVAL)` if the index is not valid.
+    /// - `Err(ErrorCode::ALREADY)` if the key is already selected.
+    fn select_key(&self, index: usize) -> Result<(), ErrorCode>;
+
+    fn set_client(&self, client: &'a dyn SelectKeyClient);
+}
+
+/// Client for setting keys.
+pub trait SetKeyClient<const KL: usize> {
+    /// Called when the key has been set.
+    ///
+    /// Returns the previous key if one was set.
+    ///
+    /// ### `error`:
+    ///
+    /// - `Ok(())`: The key was set successfully.
+    /// - `Err(())`: The key was not set successfully.
+    ///   - `ErrorCode::FAIL`: The key could not be set.
+    fn set_key_done(
+        &self,
+        previous_key: Option<&'static mut [u8; KL]>,
+        error: Result<(), ErrorCode>,
+    );
+}
+
+/// Interface for setting keys by a slice.
+///
+/// `KL` is the length of the keys.
+///
+/// Implementers must be able to store keys from a slice. This is most commonly
+/// used for implementations that hold keys in memory. However, this interface
+/// is asynchronous as keys may be stored in external storage or an external
+/// chip and require an asynchronous operations.
+pub trait SetKey<'a, const KL: usize> {
+    /// Set the current key.
+    ///
+    /// This is asynchronous. If there is an existing key that key will be
+    /// returned in the `set_key_done()` callback.
```

Comment:
Normally this idiom makes a lot of sense, but for keys in particular, should we tightly couple returning the currently set key with setting a new one?

Maybe the answer is that this is low-level enough that something that can set a key can read it, but I wonder if there isn't still value with crypto things to be a bit more explicit about when keys are being read -- it's not immediately obvious that setting a key would read out a key!

The concrete suggestion being something like a "extract_key" method if you do need to read out the old value. Perhaps coupled with a "drop_key" method to free the slot if you don't, and semantics that a slot has to be empty in order for a key to be set.